### ISSUE TITLE
galera/src/trx_handle.cpp:52:1: error: 'std::ostream& galera::operator<<(std::ostream&, const galera::TrxHandle&)' has not been declared within galera

### ISSUE THREAD

**Issue by philip-galera** [codership/galera#421]
Quoting from Otto:

latest galera does not build on the Debian
reproducable build system. It fails with:

scons: **\* [galera/src/trx_handle.o] Error 1
galera/src/trx_handle.cpp:52:1: error: 'std::ostream&
galera::operator<<(std::ostream&, const galera::TrxHandle&)' has not
been declared within galera [-Werror]
 galera::operator<<(std::ostream& os, const TrxHandle& th)
 ^~~~~~
In file included from galera/src/trx_handle.cpp:5:0:
galera/src/trx_handle.hpp:755:30: note: only here as a friend
         friend std::ostream& operator<<(std::ostream& os, const
TrxHandle& trx);
                              ^~~~~~~~

Full log at https://tests.reproducible-builds.org/debian/rbuild/unstable/amd64/galera-3_25.3.17-1.rbuild.log
and full details of galera-3 reproducible builds at
https://tests.reproducible-builds.org/debian/rb-pkg/unstable/amd64/galera-3.html

**Comment by philip-galera** [comment]
@ottok I filed your report about compilation failure here .

**Comment by ayurchen** [comment]
@ottok Precisely this issue was supposed to be fixed in https://github.com/codership/galera/pull/405/commits/5c656b1805e4e5636ecf3dd9191869c9eedf42be and is a part of 3.17 release. The fix is straightforward and operator<< is declared just few lines below. Seeing this I suspect that wrong sources were fetched for compilation. What command is used to fetch the sources for compilation by the build machine? When checking out 'release_25.3.17' tag I get this declaration.

**Comment by ottok** [comment]
The source is here: https://anonscm.debian.org/cgit/pkg-mysql/galera-3.git/

Tag https://anonscm.debian.org/cgit/pkg-mysql/galera-3.git/tag/?h=debian/25.3.17-1 points to what was uploaded to Debian and tag https://anonscm.debian.org/cgit/pkg-mysql/galera-3.git/tag/?h=upstream/25.3.17 points to what was imported from upstream tar.gz contents.

**Comment by ottok** [comment]
Slightly easier navigation at the mirror of that official packaging git repo at https://github.com/ottok/galera-3

**Comment by ayurchen** [comment]
@ottok Well, the sources that I could find through the links show that the fix is there. So it is quite strange - if it were these sources that were compiled. There is still some doubt about it, as compiler flags show reference to 3.16:

```
-g -O2 -fdebug-prefix-map=/build/galera-3-25.3.16=. -fstack-protector-strong -Wformat -Werror=format-security -std=c99 -fno-strict-aliasing -pipe
 -g -O3 -DNDEBUG -m64 -pthread -Wall -Wextra -Wno-unused-parameter -Werror 
-Wdate-time -D_FORTIFY_SOURCE=2 -D_XOPEN_SOURCE=600 -DHAVE_COMMON_H -DGALERA_USE_GU_NETWORK -DHAVE_BYTESWAP_H -DHAVE_ENDIAN_H -DHAVE_EXECINFO_H -DHAVE_BOOST_SHARED_PTR_HPP -DHAVE_TR1_UNORDERED_MAP -DBOOST_DATE_TIME_POSIX_TIME_STD_CONFIG=1 -DHAVE_SYSTEM_ASIO -DHAVE_ASIO_HPP -DHAVE_ASIO_SSL_HPP -DGCS_USE_GCOMM -DGALERA_LOG_H_ENABLE_CXX -Wno-variadic-macros -Wno-old-style-cast -Wno-missing-field-initializers -Wno-effc++
-g -O2 -fdebug-prefix-map=/build/galera-3-25.3.16=. -fstack-protector-strong -Wformat -Werror=format-security -Wno-long-long -Wno-deprecated -ansi -pipe -Weffc++ -Wold-style-cast
scons: done reading SConscript files.
```

If the above is not relevant, then I'll have to reproduce the issue somehow. Is there a way to see which compiler was used for compilation?

**Comment by ottok** [comment]
@ayurchen You are right, the log does look weird. It starts off with the .17 sources until suddently: 

```
I: Copying source file
I: copying [galera-3_25.3.16-1.dsc]
I: copying [./galera-3_25.3.16.orig.tar.gz]
I: copying [./galera-3_25.3.16-1.debian.tar.xz]
I: Extracting source
gpgv: unknown type of key resource 'trustedkeys.kbx'
gpgv: keyblock resource '/root/.gnupg/trustedkeys.kbx': General error
gpgv: Signature made Mon May 30 21:18:27 2016 GMT+12 using RSA key ID 
gpgv: Can't check signature: No public key
dpkg-source: warning: failed to verify signature on ./galera-3_25.3.16-1.dsc
dpkg-source: info: extracting galera-3 in galera-3-25.3.16
dpkg-source: info: unpacking galera-3_25.3.16.orig.tar.gz
dpkg-source: info: unpacking galera-3_25.3.16-1.debian.tar.xz
```

I will report this to Debian RB team as I suspect the build is somehow broken.

**Comment by ottok** [comment]
Ok, so for amd64 and arm it was a false positive. Turns out there was a bug in the Debian reproducible build source download logic.

However i386 still fails, what do you think this is: https://tests.reproducible-builds.org/debian/rb-pkg/unstable/i386/galera-3.html ?

**Comment by temeo** [comment]
I believe i386 issue is https://github.com/codership/galera/issues/388, IIRC it is a result of having 32 bit buildroot on 64 bit kernel.

**Comment by temeo** [comment]
Build log https://tests.reproducible-builds.org/debian/rb-pkg/unstable/i386/galera-3.html shows that `-mx32` is passed to compiler.
